File: iosApp/Sources/ClipList/DrawerList.swift
import SwiftUI

/// The drawer list's shared metrics, from the `Rally Drawer` mock.
///
/// The list is an inset one now, not a full-bleed one: rows are cards sitting on
/// the panel with a margin either side, separated by a gap rather than by a
/// divider. Everything that renders a row in the drawer - the "On this phone"
/// card in LocalVideoSection.swift and both match rows in ClipListView.swift -
/// reads its numbers from here, so the two files cannot drift a point apart.
///
/// Mirrors androidApp's DrawerList.kt number for number.
enum DrawerList {
    /// The panel's own gutter. The mock's `margin: … 24px`.
    static let sideMargin: CGFloat = 24
    static let rowPaddingH: CGFloat = 16
    static let rowPaddingV: CGFloat = 14
    /// Thumbnail to text, and text to the trailing control.
    static let gap: CGFloat = 12
    /// Between two rows. Replaces the separator the old full-bleed list drew.
    static let rowGap: CGFloat = 6
    /// Above a section label. `rowGap` adds to this between two sections.
    static let sectionGap: CGFloat = 22
    /// A section label to its first row. `rowGap` adds to this.
    static let labelGap: CGFloat = 4
    /// The list's own top inset. On Android the LazyColumn carries it as
    /// contentPadding and every label carries `sectionGap`, which puts the first
    /// label 26dp below the drawer header; `firstSectionGap` is that same sum,
    /// stated once, because a SwiftUI row carries its whole inset itself.
    static let listTopInset: CGFloat = 4
    /// Above the FIRST label, measured from the drawer's header. Unlike a later
    /// label there is no row above it to contribute `rowGap`.
    static let firstSectionGap: CGFloat = sectionGap + listTopInset
    static let thumbWidth: CGFloat = 64
    static let thumbHeight: CGFloat = 40
}

extension View {
    /// Strips the styling `List` gives a row so an inset card can be drawn in
    /// its place: no separator, no system row fill over the drawer's own panel,
    /// and the row's margin and gap stated as insets rather than as spacing
    /// between rows (`listRowSpacing` is a `List` initialiser argument, and this
    /// list needs different gaps above a section label and above a row).
    ///
    /// [top] carries whatever extra the caller needs above this row - a section
    /// label asks for `DrawerList.sectionGap`.
    func drawerListRow(top: CGFloat = 0) -> some View {
        self
            .listRowInsets(EdgeInsets(
                top: top,
                leading: DrawerList.sideMargin,
                bottom: DrawerList.rowGap,
                trailing: DrawerList.sideMargin
            ))
            .listRowBackground(Color.clear)
            .listRowSeparator(.hidden)
    }
}

/// A drawer section's label: 12pt, secondary, sentence case.
///
/// Deliberately not `Shuttl.sectionLabel`, the uppercase tracked `labelSmall`
/// the app's older lists use: the mocks for this redesign carry no uppercase and
/// no positive tracking anywhere. AnalyticsListView.swift has since moved to
/// this form and states its own spacing, so it draws the label itself rather
/// than calling this; the two want one shared component. ShareSheetView.swift is
/// the last screen still on the older form.
struct DrawerSectionLabel: View {
    let text: String

    init(_ text: String) { self.text = text }

    var body: some View {
        Text(text)
            .shuttlType(ShuttlType.bodySmall)
            .foregroundStyle(Shuttl.textSecondary)
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(.bottom, DrawerList.labelGap)
    }
}

File: androidApp/src/main/java/com/badmintontracker/android/cliplist/DrawerList.kt
package com.badmintontracker.android.cliplist

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Shape
import androidx.compose.ui.unit.dp
import com.badmintontracker.android.ui.components.SwipeToRemoveRow

/**
 * The drawer list's shared metrics, from the `Rally Drawer` mock.
 *
 * The list is an inset one now, not a full-bleed one: rows are cards sitting on
 * the panel with a margin either side, separated by a gap rather than by a
 * divider. Everything that renders a row in the drawer - the "On this phone"
 * card in LocalVideoSection.kt and both match rows in ClipListScreen.kt - reads
 * its numbers from here, so the two files cannot drift a dp apart.
 *
 * Mirrors iosApp's DrawerList.swift number for number.
 */
internal object DrawerList {
    /** The panel's own gutter. The mock's `margin: … 24px`. */
    val sideMargin = 24.dp
    val rowPaddingH = 16.dp
    val rowPaddingV = 14.dp
    /** Thumbnail to text, and text to the trailing control. */
    val gap = 12.dp
    /** Between two rows. Replaces the divider the old full-bleed list drew. */
    val rowGap = 6.dp
    /** Above a section label. [rowGap] adds to this between two sections. */
    val sectionGap = 22.dp
    /** A section label to its first row. [rowGap] adds to this. */
    val labelGap = 4.dp
    /** The list's own top inset, so the first label clears the header by 26dp. */
    val listTopInset = 4.dp
    val thumbWidth = 64.dp
    val thumbHeight = 40.dp
}

/**
 * A drawer section's label: 12sp, secondary, sentence case.
 *
 * Deliberately not the uppercase tracked `labelSmall` the app's older lists
 * use: the mocks for this redesign carry no uppercase and no positive tracking
 * anywhere. AnalyticsScreen.kt's SectionHeader has since moved to this form and
 * states its own spacing, so it draws the label itself rather than calling this;
 * the two want one shared component.
 */
@Composable
internal fun DrawerSectionLabel(text: String) {
    Text(
        text = text,
        style = MaterialTheme.typography.bodySmall,
        color = MaterialTheme.colorScheme.onSurfaceVariant,
        modifier = Modifier.padding(
            start = DrawerList.sideMargin,
            end = DrawerList.sideMargin,
            // Carried by the label rather than by the list, so the gap above a
            // section is the same whether it is the first one or the third.
            // The list's own top inset makes up the difference for the first;
            // for the rest the [DrawerList.rowGap] between items adds to this.
            top = DrawerList.sectionGap,
            bottom = DrawerList.labelGap,
        ),
    )
}

/**
 * The inset, rounded box every drawer row sits in.
 *
 * Exists because a rounded inset row and [SwipeToRemoveRow] have to agree on
 * three things at once, and getting any of them wrong is visible: the margin
 * (or the red reveal bleeds past the card to the panel edge), the shape (or
 * square corners poke out from behind the rounded silhouette mid-swipe), and
 * the fill behind the content (or the reveal shows through the row before the
 * user has swiped). Stating them once, here, is what keeps a swipeable row and
 * a non-swipeable one identical.
 *
 * [swipeLabel] null renders the same box without the gesture - the mid-pipeline
 * rows that must not be removable.
 */
@Composable
internal fun DrawerRowContainer(
    shape: Shape,
    fill: Color,
    swipeLabel: String? = null,
    onSwiped: () -> Boolean = { false },
    content: @Composable () -> Unit,
) {
    val inset = Modifier.padding(horizontal = DrawerList.sideMargin)
    if (swipeLabel != null) {
        SwipeToRemoveRow(
            label = swipeLabel,
            onSwiped = onSwiped,
            modifier = inset,
            shape = shape,
            contentBackground = fill,
            content = content,
        )
    } else {
        Box(inset.clip(shape).background(fill)) { content() }
    }
}
